File: python/Brunel/_keydate.py

__all__ = ["KeyDate"]


def _setState(state, val, default=None):
    try:
        result = state[val]
        if result:
            return result
        else:
            return default
    except Exception:
        return default


class KeyDate:
    """Holds information about a key date in the project"""
    def __init__(self, props=None, getHook=None):
        self._getHook = getHook

        self.state = {
            "name": None,
            "id": None,
            "description": None,
        }

        self.setState(props)

    def __str__(self):
        return f"KeyDate({self.getName()})"

    def __repr__(self):
        return self.__str__()

    def getID(self):
        return self.state["id"]

    def getName(self):
        return self.state["name"]

    def getDescription(self):
        return self.state["description"]

    def setState(self, state):
        if not state:
            return

        self.state["name"] = _setState(state, "name")
        self.state["id"] = _setState(state, "id")
        self.state["description"] = _setState(state, "description")

        if self.state["name"] == "None" or self.state["name"] is None:
            raise ValueError(f"No name for {self}?")

    def toDry(self):
        return self.state

    @staticmethod
    def unDry(value):
        return KeyDate(value)

    @staticmethod
    def load(data):
        return KeyDate(data)

File: src/model/KeyDate.js
import Dry from "json-dry";
import lodash from "lodash";

import { ValueError } from "./Errors";

function setState(val, def = null) {
  if (val) {
    return val;
  } else {
    return def;
  }
}

class KeyDate {
  constructor(props) {
    this.state = {
      name: null,
      id: null,
      description: null,
      month: null,
    };

    this.setState(props);

    this._getHook = null;
    this._isAKeyDateObject = true;
  }

  static clone(item) {
    let c = new KeyDate();
    c._getHook = item._getHook;
    c.state = lodash.cloneDeep(item.state);
    return c;
  }

  getID() {
    return this.state.id;
  }

  setState(state) {
    if (state) {
      this.state.name = setState(state.name);
      this.state.id = setState(state.id);
      this.state.description = setState(state.description, "");
      this.state.month = setState(state.month, null);

      if (!this.state.name) {
        throw new ValueError("You cannot have a KeyDate without a name");
      }
    }
  }

  _updateHooks(hook) {
    this._getHook = hook;
  }

  merge() {
    return this;
  }

  toString() {
    return `KeyDate(${this.getName()})`;
  }

  getName() {
    return this.state.name;
  }

  getDescription() {
    return this.state.description;
  }

  getMonth() {
    return this.state.month;
  }

 
  toDry() {
    return { value: this.state };
  }
}

KeyDate.unDry = function (value) {
  return new KeyDate(value);
};

Dry.registerClass("KeyDate", KeyDate);

export default KeyDate;
